cpp/osh_bool_stat.h

```diff
@@ -18,26 +19,101 @@ bool isatty(Str* fd_str, word_t* blame_word);
 inline bool DoUnaryOp(Id_t op_id, Str* s) {
   mylib::Str0 path(s);
 
-  // TODO: also call lstat(), etc.
-  struct stat st;
-  if (stat(path.Get(), &st) < 0) {
-    return false;
-  }
-  auto mode = st.st_mode;
-
-  switch (op_id) {
-  // synonyms for existence
-  case Id::BoolUnary_a:
-  case Id::BoolUnary_e:
-    return true;
+  const char *zPath = path.Get();
 
-  case Id::BoolUnary_f:
-    return S_ISREG(mode);
+  if (op_id == Id::BoolUnary_h ||
+      op_id == Id::BoolUnary_L )
+  {
+    struct stat st;
+    if (lstat(zPath, &st) < 0) {
+      return false;
+    }
 
-  case Id::BoolUnary_k:
-    return (mode & S_ISVTX) != 0;
+    return S_ISLNK(st.st_mode);
+  }
+  else
+  {
+    struct stat st;
+    if (stat(zPath, &st) < 0) {
+      return false;
+    }
+
+    auto mode = st.st_mode;
+
+    switch (op_id) {
+    // synonyms for existence
+    case Id::BoolUnary_a:
+    case Id::BoolUnary_e:
+      return true;
+
+    case Id::BoolUnary_s:
+      return st.st_size != 0;
+
+    case Id::BoolUnary_d:
+      return S_ISDIR(mode);
+
+    case Id::BoolUnary_f:
+      return S_ISREG(mode);
+
+    case Id::BoolUnary_k:
+      return (mode & S_ISVTX) != 0;
+
+    case Id::BoolUnary_p:
+      return S_ISFIFO(mode);
+
+    case Id::BoolUnary_O:
+      return st.st_uid == geteuid();
+
+    case Id::BoolUnary_G:
+      return st.st_gid == getegid();
+
+    case Id::BoolUnary_u:
+      return mode & S_ISUID;
+
+    case Id::BoolUnary_g:
+      return mode & S_ISGID;
+
+
+    // NOTE(Jesse): This implementation MAY have a bug.  On my system (Ubuntu
+    // 20.04) it returns a correct result if the user is root (elevated with
+    // sudo) and no execute bits are set for a file.
```

Comment:
Oh OK maybe the workarounds are only for old non-POSIX  versions of faccessat?   In which case this could be correct!  

thanks for documenting it either way